numpy/core/src/umath/ufunc_object.c

```diff
@@ -924,21 +935,60 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
     PyObject *obj;
 
     /* Convert and fill in input arguments */
+    npy_bool all_scalar = NPY_TRUE;
+    npy_bool any_scalar = NPY_FALSE;
+    npy_bool all_legacy = NPY_TRUE;
     for (int i = 0; i < nin; i++) {
         obj = PyTuple_GET_ITEM(full_args.in, i);
 
         if (PyArray_Check(obj)) {
-            PyArrayObject *obj_a = (PyArrayObject *)obj;
-            out_op[i] = (PyArrayObject *)PyArray_FromArray(obj_a, NULL, 0);
+            out_op[i] = (PyArrayObject *)obj;
+            Py_INCREF(out_op[i]);
         }
         else {
-            out_op[i] = (PyArrayObject *)PyArray_FromAny(obj,
-                                    NULL, 0, 0, 0, NULL);
+            /* Convert the input to an array and check for special cases */
+            out_op[i] = (PyArrayObject *)PyArray_FromAny(obj, NULL, 0, 0, 0, NULL);
+            if (out_op[i] == NULL) {
```

Comment:
This is a bit awkward. The argument is converted to an array, but if it is one of the special objects (python long, float, complex) the `out_op_DTypes` is marked with the type, and later on, value-based promotion is used. Wouldn't it be better to avoid the conversion here, since the value-based promotion can use the original object anyway?

Reply:
After trying to figure out a way to make this nicely, I opted to just drop it for now with a huge comment.  It probably makes more sense to defer this till after we attempt getting rid of value-based casting for good (which I assume I can try very soon if this moves well!)

That said, it is probably important to get this right (or warn about it) when we make DTypes public, so added it to the issue.